lang/java/thrift/src/test/java/org/apache/avro/thrift/TestThrift.java

```diff
@@ -93,4 +94,25 @@ private void check(Test test) throws Exception {
     assertEquals(test, o);
 
   }
+
+  @org.junit.Test
+  public void complexStructTest() {
+    Schema schema = ThriftData.get().getSchema(Test.class);
+
+    Schema intOptional = Schema.createUnion(Schema.create(Schema.Type.NULL), Schema.create(Schema.Type.INT));
+    Schema stringSchema = Schema.create(Schema.Type.STRING);
+    stringSchema.addProp("avro.java.string", "String");
+    stringSchema = Schema.createUnion(Schema.create(Schema.Type.NULL), stringSchema);
+
+    Schema stringOptionalDefaultValue = Schema.create(Schema.Type.STRING);
+    stringOptionalDefaultValue.addProp("avro.java.string", "String");
+    stringOptionalDefaultValue = Schema.createUnion(stringOptionalDefaultValue, Schema.create(Schema.Type.NULL));
```

Comment:
Why is the order here different than the for the `stringSchema`? Also, this PR doesn't address the problem of not caring about the default value in the thrift does it?

Reply:
The order of two schemas for different strings is different because one of them has default value while another not. So, as avro spec recommends, in union type the type of default value should be the first -> if optional field has no default value then the default value is null.

Yes, this PR is not about default values because i've faced some problems with it: currently it is not possible to get default values for primitive types (more precisely, it's possible but you do not know for sure is it a default value in thrift file or just a variable default value) and values for non-primitive types (such as map, list, set, record and unions) are not mapped correctly to avro default values.

As i could not solve the problem with default values i suggest to consider only problem with order of union type (<NULL, TYPE> or <TYPE, NULL>) for optional fields and add <NULL_DEFAULT_VALUE> for optional types to preserve compatibility.